File: templates/python-tests/src/ingest/kafka_tests.py
from pydantic import BaseModel
from moose_lib import Key, OlapTable, OlapConfig, Stream, IngestApi
from moose_lib.blocks import KafkaEngine
from moose_lib.dmv2 import MaterializedView, MaterializedViewOptions, IngestConfigWithDestination


class KafkaTestEvent(BaseModel):
    event_id: Key[str]
    user_id: str
    event_type: str
    amount: float
    timestamp: int


kafka_test_input_stream = Stream[KafkaTestEvent]("kafka_test_input")

kafka_test_ingest_api = IngestApi[KafkaTestEvent](
    "kafka-test",
    IngestConfigWithDestination(destination=kafka_test_input_stream)
)

kafka_test_source_table = OlapTable[KafkaTestEvent](
    "kafka_test_source",
    OlapConfig(
        engine=KafkaEngine(
            broker_list="redpanda:9092",
            topic_list="kafka_test_input",
            group_name="e2e_kafka_test_consumer_py",
            format="JSONEachRow"
        ),
        settings={
            "kafka_num_consumers": "1",
        },
    ),
)

kafka_test_mv_query = """
SELECT
    event_id,
    user_id,
    event_type,
    amount,
    timestamp
FROM kafka_test_source
"""

kafka_test_mv = MaterializedView[KafkaTestEvent](
    MaterializedViewOptions(
        select_statement=kafka_test_mv_query,
        select_tables=[kafka_test_source_table],
        table_name="kafka_test_dest",
        materialized_view_name="kafka_test_dest_mv",
        order_by_fields=["event_id", "timestamp"],
    )
)

File: templates/typescript-tests/src/ingest/kafkaTests.ts
import {
  Stream,
  IngestApi,
  OlapTable,
  MaterializedView,
  ClickHouseEngines,
  Key,
  sql,
} from "@514labs/moose-lib";

export interface KafkaTestEvent {
  eventId: Key<string>;
  userId: string;
  eventType: string;
  amount: number;
  timestamp: number;
}

export const kafkaTestInputStream = new Stream<KafkaTestEvent>(
  "KafkaTestInput",
);

export const kafkaTestIngestApi = new IngestApi<KafkaTestEvent>("kafka-test", {
  destination: kafkaTestInputStream,
});

export const KafkaTestSourceTable = new OlapTable<KafkaTestEvent>(
  "KafkaTestSource",
  {
    engine: ClickHouseEngines.Kafka,
    brokerList: "redpanda:9092",
    topicList: "KafkaTestInput",
    groupName: "e2e_kafka_test_consumer",
    format: "JSONEachRow",
    settings: {
      kafka_num_consumers: "1",
    },
  },
);

const kafkaSourceColumns = KafkaTestSourceTable.columns;

export const KafkaTestMV = new MaterializedView<KafkaTestEvent>({
  tableName: "KafkaTestDest",
  materializedViewName: "KafkaTestDest_MV",
  orderByFields: ["eventId", "timestamp"],
  selectStatement: sql`
    SELECT
      ${kafkaSourceColumns.eventId} as eventId,
      ${kafkaSourceColumns.userId} as userId,
      ${kafkaSourceColumns.eventType} as eventType,
      ${kafkaSourceColumns.amount} as amount,
      ${kafkaSourceColumns.timestamp} as timestamp
    FROM ${KafkaTestSourceTable}
  `,
  selectTables: [KafkaTestSourceTable],
});
